src/miral/application_selector.cpp

```diff
@@ -0,0 +1,130 @@
+    if (!application)
+    {
+        return nullptr;
+    }
+
+    selected = tools.info_for(application);
+    reverse = in_reverse;
+    is_started = true;
+    raise_next();
+    return selected.application();
+}
+
+auto ApplicationSelector::next() -> Application
+{
+    if (!is_active())
+    {
+        mir::log(mir::logging::Severity::warning, MIR_LOG_COMPONENT,
+                 "Cannot call ApplicationSelector::next when the ApplicationSelector is not active. Call ApplicationSelector::start first.");
+        return nullptr;
+    }
+
+    raise_next();
+    return selected.application();
+}
+
+auto ApplicationSelector::complete() -> Application
+{
+    if (!is_active())
+    {
+        mir::log(mir::logging::Severity::warning, MIR_LOG_COMPONENT,
+                 "Cannot call ApplicationSelector::stop when the ApplicationSelector is not active. Call ApplicationSelector::start first.");
+        return nullptr;
+    }
+
+    is_started = false;
+    if (tools.can_focus_application(selected.application())) {
+        // Note: The null checks here are mostly to accommodate the tests, however they are a good idea.
+        auto application = selected.application();
+        auto surface = application ? application->default_surface() : nullptr;
+        if (surface)
+        {
+            auto window = tools.info_for(surface).window();
+            tools.select_active_window(window);
+        }
+        else
+        {
+            mir::log(mir::logging::Severity::warning, MIR_LOG_COMPONENT,
+                     "ApplicationSelector::complete: Failed to select the active window.");
+        }
+    }
+
+    return selected.application();
+}
+
+auto ApplicationSelector::is_active() -> bool
+{
+    return is_started;
+}
+
+void ApplicationSelector::raise_next()
+{
+    // First, find a suitable application in the list based off of the direction.
+    // If we encounter the presently selected application again, that means
+    // we have gone all the way around the list, so we can "break".
+    auto selected_app = selected.application();
+    auto app_info = selected;
+    do {
+        app_info = reverse ? tools.get_previous_application_info(app_info)
+            : tools.get_next_application_info(app_info);
+        if (app_info.application() == selected_app) break;
+    } while (!tools.can_focus_application(app_info.application()));
+
+    // Next, raise the window, but note that we do not actually focus it.
+    // Note: The null checks here are mostly to accommodate the tests, however they are a good idea.
+    auto application = app_info.application();
+    auto surface = application ? application->default_surface() : nullptr;
+    if (surface)
+    {
+        auto window = tools.info_for(surface).window();
+        tools.raise_tree(window);
+    }
+    else
+    {
+        mir::log(mir::logging::Severity::warning, MIR_LOG_COMPONENT,
+                 "ApplicationSelector::raise_next: Failed to raise the newly selected window.");
```

Comment:
There's a `log_warning()` convenience wrapper...

```suggestion
        mir::log_warning("ApplicationSelector::raise_next: Failed to raise the newly selected window.");
```